yamux/src/connection/stream.rs

```diff
@@ -316,37 +317,47 @@ impl AsyncRead for Stream {
         }
 
         // Copy data from stream buffer.
-        let mut shared = self.shared();
         let mut n = 0;
-        while let Some(chunk) = shared.buffer.front_mut() {
-            if chunk.is_empty() {
-                shared.buffer.pop();
-                continue;
+        let can_read = self.shared.with_mut(|inner| {
```

Comment:
Can't we just return here directly the `Poll` result?
That way we could move the whole bottom part of the function as it was inside this closure.

Reply:
It's now using simple `lock` instead of `with_mut`, so no need for this anymore.